### PR TITLE
refactor: inter-chain-token service events and errors [AXE-2064]

### PR DESCRIPTION
**Pull request by blockchainguyy**
Update errors and events as suggested [here](https://www.notion.so/bright-ambert-2bd/Axelar-ITS-Review-and-Monitoring-Plan-5319e0c5bcec4312a673e6465fba3687)
Also fixed the tests due to these changes

### CODE REVIEW

**Review by Foivos — APPROVED** [review]

**Review by deanamiel — APPROVED** [review]

File: contracts/interfaces/IFlowLimit.sol [hunk 1]
```diff
@@ -3,9 +3,9 @@
 pragma solidity ^0.8.0;
 
 interface IFlowLimit {
-    error FlowLimitExceeded();
+    error FlowLimitExceeded(uint256 limit, uint256 flowAmount);
```
Comment (milapsheth) [inline contracts/interfaces/IFlowLimit.sol:6]: shall we include the token id/token address here as well?

File: contracts/interfaces/IFlowLimit.sol [hunk 2]
```diff
@@ -3,9 +3,9 @@
...
 
-    event FlowLimitSet(uint256 flowLimit);
+    event FlowLimitSet(address operator, uint256 flowLimit);
```
Comment (milapsheth) [inline contracts/interfaces/IFlowLimit.sol:8]: same. or do we expect to go from token manager -> token id off chain

File: contracts/interfaces/IDistributable.sol [hunk 3]
```diff
@@ -3,10 +3,10 @@
 pragma solidity ^0.8.0;
 
 interface IDistributable {
-    error NotDistributor();
-    error NotProposedDistributor();
+    error NotDistributor(address caller);
```
Comment (milapsheth) [inline contracts/interfaces/IDistributable.sol:6]: can you revert these changes? the new access control will deprecate these errors so less conflicts for the next PR

File: contracts/utils/FlowLimit.sol [hunk 4]
```diff
@@ -99,8 +99,9 @@ contract FlowLimit is IFlowLimit {
             flowToCompare := sload(slotToCompare)
         }
 
-        if (flowToAdd + flowAmount > flowToCompare + flowLimit) revert FlowLimitExceeded();
-        if (flowAmount > flowLimit) revert FlowLimitExceeded();
+        if (flowToAdd + flowAmount > flowToCompare + flowLimit)
+            revert FlowLimitExceeded((flowToCompare + flowLimit), flowToAdd + flowAmount);
```
Comment (milapsheth) [inline contracts/utils/FlowLimit.sol:103]: store the values in temp variables perhaps to avoid duplication?

**Review by Foivos — APPROVED** [review]

File: contracts/interchain-token-service/InterchainTokenService.sol [hunk 5]
```diff
@@ -129,7 +129,8 @@ contract InterchainTokenService is
      * @param tokenId the `tokenId` of the TokenManager trying to perform the call.
      */
     modifier onlyTokenManager(bytes32 tokenId) {
-        if (msg.sender != getTokenManagerAddress(tokenId)) revert NotTokenManager();
+        address tokenManager = getTokenManagerAddress(tokenId);
+        if (msg.sender != tokenManager) revert NotTokenManager(msg.sender, tokenManager);
         _;
```
Comment (milapsheth) [inline contracts/interchain-token-service/InterchainTokenService.sol:134]:
```suggestion

        _;
```

File: contracts/interchain-token-service/InterchainTokenService.sol [hunk 6]
```diff
@@ -278,10 +279,13 @@ contract InterchainTokenService is
...
-        address tokenAddress = getValidTokenManagerAddress(tokenId);
-        tokenAddress = ITokenManager(tokenAddress).tokenAddress();
-
-        if (getCanonicalTokenId(tokenAddress) != tokenId) revert NotCanonicalTokenManager();
+        address tokenAddress;
+        {
+            tokenAddress = getValidTokenManagerAddress(tokenId);
+            tokenAddress = ITokenManager(tokenAddress).tokenAddress();
+            bytes32 canonicalTokenId = getCanonicalTokenId(tokenAddress);
+            if (canonicalTokenId != tokenId) revert InvalidCanonicalTokenId(canonicalTokenId);
```
Comment (milapsheth) [inline contracts/interchain-token-service/InterchainTokenService.sol:287]:
```suggestion

            if (canonicalTokenId != tokenId) revert InvalidCanonicalTokenId(canonicalTokenId);
```

File: contracts/interchain-token-service/InterchainTokenService.sol [hunk 7]
```diff
@@ -780,14 +785,18 @@ contract InterchainTokenService is
...
             abi.encodeWithSelector(ITokenManagerDeployer.deployTokenManager.selector, tokenId, tokenManagerType, params)
         );
-        if (!success) revert TokenManagerDeploymentFailed();
+        if (!success) revert TokenManagerDeploymentFailed(returnData);
+
+        address tokenManager;
+        assembly {
+            tokenManager := mload(add(returnData, 0x20))
+        }
+        // slither-disable-next-line reentrancy-events
```
Comment (milapsheth) [inline contracts/interchain-token-service/InterchainTokenService.sol:798]:
```suggestion

        // slither-disable-next-line reentrancy-events
```

File: contracts/interchain-token-service/InterchainTokenService.sol [hunk 8]
```diff
@@ -838,8 +845,14 @@ contract InterchainTokenService is
             )
         );
         if (!success) {
-            revert StandardizedTokenDeploymentFailed();
+            revert StandardizedTokenDeploymentFailed(returnData);
         }
+        address tokenAddress;
```
Comment (milapsheth) [inline contracts/interchain-token-service/InterchainTokenService.sol:850]:
```suggestion

        address tokenAddress;
```

File: contracts/interchain-token-service/InterchainTokenService.sol [hunk 9]
```diff
@@ -838,8 +845,14 @@ contract InterchainTokenService is
...
+        assembly {
+            tokenAddress := mload(add(returnData, 0x20))
+        }
+        // slither-disable-next-line reentrancy-events
```
Comment (milapsheth) [inline contracts/interchain-token-service/InterchainTokenService.sol:854]:
```suggestion

        // slither-disable-next-line reentrancy-events
```

File: contracts/utils/StandardizedTokenDeployer.sol [hunk 10]
```diff
@@ -45,12 +45,12 @@ contract StandardizedTokenDeployer is IStandardizedTokenDeployer, Create3 {
         uint8 decimals,
         uint256 mintAmount,
         address mintTo
-    ) external payable {
+    ) external payable returns (address tokenAddress) {
```
Comment (milapsheth) [inline contracts/utils/StandardizedTokenDeployer.sol:48]: update docstring with return value

File: contracts/utils/TokenManagerDeployer.sol [hunk 11]
```diff
@@ -18,14 +18,19 @@ contract TokenManagerDeployer is ITokenManagerDeployer, Create3 {
...
      * @param params Additional parameters used in the setup of the token manager
+     * @return tokenManagerAddress The address of the deployed tokenManager
      */
     // slither-disable-next-line locked-ether
-    function deployTokenManager(bytes32 tokenId, uint256 implementationType, bytes calldata params) external payable {
+    function deployTokenManager(
+        bytes32 tokenId,
+        uint256 implementationType,
+        bytes calldata params
+    ) external payable returns (address tokenManagerAddress) {
```
Comment (milapsheth) [inline contracts/utils/TokenManagerDeployer.sol:28]:
```suggestion
    ) external payable returns (address tokenManager) {
```
since address is redundant

File: contracts/proxies/TokenManagerProxy.sol [hunk 12]
```diff
@@ -28,8 +28,8 @@ contract TokenManagerProxy is ITokenManagerProxy {
         tokenId = tokenId_;
         address impl = _getImplementation(IInterchainTokenService(interchainTokenServiceAddress_), implementationType_);
 
-        (bool success, ) = impl.delegatecall(abi.encodeWithSelector(TokenManagerProxy.setup.selector, params));
-        if (!success) revert SetupFailed();
+        (bool success, bytes memory error) = impl.delegatecall(abi.encodeWithSelector(TokenManagerProxy.setup.selector, params));
```
Comment (milapsheth) [inline contracts/proxies/TokenManagerProxy.sol:31]: returnData instead of error? Or change the event attribute above for consistency

File: contracts/token-manager/TokenManager.sol [hunk 13]
```diff
@@ -182,7 +182,7 @@ abstract contract TokenManager is ITokenManager, Operatable, FlowLimit, Implemen
      * @return the amount of token actually given, which will only be different than `amount` in cases where the token takes some on-transfer fee.
      */
     function giveToken(address destinationAddress, uint256 amount) external onlyService returns (uint256) {
-        _addFlowIn(amount);
+        _addFlowIn(amount, tokenId());
```
Comment (milapsheth) [inline contracts/token-manager/TokenManager.sol:185]: this is adding extra gas for every transfer. The emitter of the event will be the token manager, we can report that in the monitoring and query token id manually or in a consumer of the alert.

File: contracts/utils/ExpressCallHandler.sol [hunk 14]
```diff
@@ -36,7 +36,7 @@ contract ExpressCallHandler is IExpressCallHandler {
             prevExpressCaller := sload(slot)
         }
 
-        if (prevExpressCaller != address(0)) revert AlreadyExpressCalled();
+        if (prevExpressCaller != address(0)) revert AlreadyExpressCalled(prevExpressCaller);
```
Comment (milapsheth) [inline contracts/utils/ExpressCallHandler.sol:39]: also add `expressCaller` as an arg?

**Review by deanamiel — APPROVED** [review]

**Review by milapsheth — APPROVED** [review]
